linalg/tensor.hpp

```diff
@@ -0,0 +1,2051 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+/**
+ * @file tensor.hpp
+ *
+ * @brief Implementation of the tensor class
+ */
+
+#ifndef MFEM_INTERNAL_TENSOR_HPP
+#define MFEM_INTERNAL_TENSOR_HPP
+
+#include "dual.hpp"
+#include <type_traits> // for std::false_type
+
+namespace mfem
+{
+namespace internal
+{
+
+#if defined(__CUDACC__)
+#if __CUDAVER__ >= 75000
+#define MFEM_SUPPRESS_NVCC_HOSTDEVICE_WARNING #pragma nv_exec_check_disable
+#else
+#define MFEM_SUPPRESS_NVCC_HOSTDEVICE_WARNING #pragma hd_warning_disable
+#endif
```

Comment:
Our minimal required CUDA version is 10.1.168 (see INSTALL file), so we can simplify this.